Apr bucket read length cut fix

Reiner030 opened the pull request:
The fixed version was also not working in my setups (behind AWS and a local setup HAProxy for testing.

I got from apr_bucket_read always  length=11 reported but real length was about 76-1184 chars.
This was also a problem with the fixed buffer size of 107 ... I found a hint for APR_BUCKET_BUFF_SIZE
in http://svn.apache.org/repos/asf/apr/apr-util/tags/0.9.6/buckets/apr_brigade.c so I used it.

In http://www.apachetutor.org/dev/brigades I found the correct usage of split/delete APR_BUCKETs and 
howto passing down the modification of the bucket to the filter chain.

Commit:
Fixing apr_bucket_read length and cut/commit behavior

1st step: fixing buffersize

(In my latest test with AWS i got up to 1184 chars into buffer at once)

Commit:
Fixing apr_bucket_read length and cut/commit behavior

2nd step: Fixing apr_bucket_remove usage
as mentioned in: http://www.apachetutor.org/dev/brigades

Commit:
Fixing apr_bucket_read length and cut/commit behavior

3rd step: Fixing apr_success => apr_pass_brigade
as mentioned in: http://www.apachetutor.org/dev/brigades

... because later filter must known about our changes ;)

Commit:
Fixing apr_bucket_read length and cut/commit behavior

0th step: Improving DEBUGging
(mmh, other improves where partly commited with other patches)

Commit:
Fixing apr_bucket_read length and cut/commit behavior

last step: Whitespace cleaning
(mmh, other cleanings where partly commited with other patches)

Reiner030 commented:
check on local haproxy:
`$ PERL_LWP_SSL_VERIFY_HOSTNAME=0 GET https://127.0.0.1:20443`

gave with `mode tcp` and backend server with option `send-proxy`:

```
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::process_connection IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (2)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in DEBUG-ED!!
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 127.0.0.1 to port=443 (1)
[Fri Jan 16 13:12:04 2015] [warn] mod_myfixip::post_read_handler IP Connection from: 127.0.0.1 [127.0.0.1] to port=443 newip=(null) (OK)
```

EDIT: I dropped accidently backend server option `send-proxy` in haproxy config.
Actually it works only on 127.0.0.1 ... I check later if there is extended haproxy config needed.

ggrandes commented:
1) You don't want parse all APR_BUCKET_BUFF_SIZE, only a MAX of 107bytes (max length of HA-Proxy header).

ggrandes commented:
2) About "only works on 127.0.0.1", I found that on loopback (same machine), tcp fragment don't occurs, always better test with 2 remote machines.

ggrandes commented:
I use this for test (http, not ssl): https://gist.github.com/ggrandes/77e28ea4213bb7b862d1

```
javac HttpFragClient.java      
java HttpFragClient <apache-ip> <apache-port>
```

ggrandes commented:
compilation give me 2 warnings:

##### missing parameters (client_ip and port), corrected:

ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, MODULE_NAME "::helocon_filter_in DEBUG: Data read from: %s to port=%d (6) - EOL CR found", _CLIENT_IP,
 c->local_addr->port);

##### too params, corrected:

ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, MODULE_NAME "::helocon_filter_in DEBUG: CMD=PROXY no dstport found.");

ggrandes commented:
3) if you use ap_pass_brigade(f->next, b) instead of APR_SUCCESS, this produces error 400 (on plain/no ssl connection).

Reiner030 commented:
1) ok, but I got with AWS ELB up to  ~1200 bytes in the first call:

```
[Thu Jan 15 09:32:12 2015] [warn] mod_myfixip::process_connection IP Connection from: 172.31.23.234 to port=443 (1)
[Thu Jan 15 09:32:12 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 172.31.23.234 to port=443 (1)
[Thu Jan 15 09:32:12 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 172.31.23.234 to port=443 (2)
[Thu Jan 15 09:32:12 2015] [warn] mod_myfixip::helocon_filter_in DEBUG: CMD=PROXY OK
[Thu Jan 15 09:32:12 2015] [warn] mod_myfixip::helocon_filter_in DEBUG: Data read from: 172.31.23.234 to port=443 (3) content=PROXY TCP4 78.52.139.18
9 10.0.6.163 62202 443\r\nng: identity\r\nHost: localhost\r\nContent-Type: application/x-www-form-urlencoded\r\nConnection: close\r\nAccept: text/html, */*
\r\nUser-Agent: Datadog Agent/5.1.1\r\n\r\n................................................................................................................
...........................................................................................................................................................
...........................................................................................................................................................
...........................................................................................................................................................
...........................................................................................................................................................
.......................................................text/html
[Thu Jan 15 09:32:12 2015] [warn] mod_myfixip::helocon_filter_in ERROR: PROXY protocol header overflow from=172.31.23.234 to port=443 length=1001
```

```
[Thu Jan 15 09:59:43 2015] [warn] mod_myfixip::process_connection IP Connection from: 172.31.23.234 to port=443 (1)
[Thu Jan 15 09:59:43 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 172.31.23.234 to port=443 (1)
[Thu Jan 15 09:59:43 2015] [warn] mod_myfixip::helocon_filter_in IP Connection from: 172.31.23.234 to port=443 (2)
[Thu Jan 15 09:59:43 2015] [warn] mod_myfixip::helocon_filter_in DEBUG: CMD=PROXY OK
[Thu Jan 15 09:59:43 2015] [warn] mod_myfixip::helocon_filter_in DEBUG: Data read from: 172.31.23.234 to port=443 (3) content=PROXY TCP4 xx.xx.176.12
 10.0.6.163 44791 443\r\n___________.......................................................................................................................
...........................................................................................................................................................
...........................................................................................................................................................
...........................................................................................................................................................
...........................................................................................................................................................
...........................................................................................................................................................
.........................................text/html
[Thu Jan 15 09:59:43 2015] [warn] mod_myfixip::helocon_filter_in ERROR: PROXY protocol header overflow from=172.31.23.234 to port=443 length=1001
```

```
[Thu Jan 15 10:12:03 2015] [notice] Apache/2.2.22 (Debian) mod_jk/1.2.37 mod_ssl/2.2.27 OpenSSL/1.0.1h configured -- resuming normal operations
[Thu Jan 15 10:12:42 2015] [warn] mod_myfixip::helocon_filter_in ERROR: PROXY protocol header overflow from=172.31.23.234 to port=443 length=1184
```

EDIT: When the  tcp fragment can occur: 
         => Then it can help to set normal max TCP size=1500 as buffer size ?

2) yes... I must then open internal fw to test from remote server ;)
    nice Java snippet... 
    I like GET/HEAD/POST perl "binaries" from package libwww-perl because they have several
    easy options for debugging.

compilation) ok... i fix it. It could be that I dropped them accidently when cleaning up my many debug lines / I tried also to avoid buffer copying because I found nice parameters for CR and LF (but this is FF char? used in above linked apr_brigade.c) in http://apr.apache.org/docs/apr/1.4/group__apr__general.html.
There are also perhaps interesting funktions:
-  APR_OFFSET
-  APR_OFFSETOF
-  APR_STRINGIFY

3) mmh, this was in my tests the main step to get to the content... otherwise I saw only debugging in error.log but no lines in my access.log and no content / and in the client a "Connection/OpenSSL handshake not possible" like message... 
=> Therefore I  came on the idea that the ProxyProtocol header is not cutted from the request (correctly) => then the SSL handshakes failed... 

I can later create some more logs with origin script in my setups.
Are there perhaps some hints/steps which where not mentioned yet in Readme ? 
E.g. I use the threaded Apache2 version and not prefork.
Also added mod_ssl and optimal also mod_spdy-beta with google-adopted mod_ssl.
( https://code.google.com/p/mod-spdy/source/browse/trunk/src/mod_spdy/mod_spdy.cc )... 
My test are made also with only "normal" mod_ssl to exlude this error source.

When checking if the module input_filter sequence is perhaps wrong I found also this very old revision
https://code.google.com/p/mod-spdy/source/browse/trunk/src/mod_spdy/mod_spdy.cc?spec=svn17&r=17

```
// mod_ssl is AP_FTYPE_CONNECTION + 5. We want to hook right before mod_ssl.
const ap_filter_type kSpdyFilterType =
    static_cast<ap_filter_type>(AP_FTYPE_CONNECTION + 4);
```

which gets lower value in compare to mod_ssl while your module uses a higher value... 

```
    /*
     * mod_ssl is AP_FTYPE_CONNECTION + 5 and mod_myfixip needs to
     * be called before mod_ssl.
     */
    ap_register_input_filter(myfixip_filter_name, helocon_filter_in, NULL, AP_FTYPE_CONNECTION + 9);
```

```
# lsb_release -d
Description:    Debian GNU/Linux 7.8 (wheezy)
# aptitude show apache2 | grep Version
Version: 2.2.22-13+deb7u4
```

ggrandes commented:
Well, seem that the main problem was refilling of buffers (ap_get_brigade). Check new version and tell me :-)
